### ISSUE TITLE
Hash Codes in Details of Bring Item

### ISSUE THREAD

**Issue by Camatobe** [vlebourl/shopping_list#2]
Hi!
Such a nice Integration!!! Love it.
I sync it every 15 min like said.

Bring is adding some weird random Hashlike Strings to every Item added (no matter if added through HA or Bring!.)
Also: Details added in Bring are overwritten by the integration. (So I Never know wich yoghurt to buy) :(

Example: 
Item in Bring:
Title: Wasser
Detail: bd53edcea01a4132a05bd6dfaf714Xxx

Item in HA: Wasser

Using HA 0.117.6 and your latest Code (you don’t have releases yet.)

Expected Behavior: Details should not be filled with random Stuff my Wife don’t know.
Also Details should not be overwritten.

**Comment by vlebourl** [comment]
Hi,
Thank you for your interest. The hash code is the way Home Assistant recognizes an element. Bring needs to be aware of that hash code in order to avoid adding an element multiple times into HA's list but rather edit an existing one, like check it for instance. So far, storing that hash in the details section is the only way I found to share that hash. I'll try to see if I can think of something else, but that probably won't be for the near future I'll afraid...

**Comment by Camatobe** [comment]
Couldn’t we find an way to use the Title and the Details in Combination as Key? 
For example using Brings Title and the Details in Brackets, so everything is visible in HA too.

Bring:
Title: Energy Drink
Detail: Green one

HA: Energy Drink (Green one)

The Combinations should be unique (Bring even melts double Names to Title + Detail: „2“

**Comment by vlebourl** [comment]
merging title and detail is a good idea, but the problem is that an item in HA is actually a dict composed of a name, an id and a boolean: 
```py
{"name": "Energy Drink", "id": "theweirdhash", "complete": false}
```
That id, randomly generated when an item is added to HA's list, is the only way to point to that given item. So I need to find a way to keep that id known to Bring, otherwise a new item with the same name but a different id will be added every time the two lists are synced.
I have an idea I need to try, but I haven't had the time yet.

**Comment by Camatobe** [comment]
Hm... Cant we use the name+Detail from bring as id (should be unique in most cases) and bypass the uuid generation? 

Also noticed this Case: 
- When adding an Item with details in Bring first
- Calling Bring-Sync Service in HA
- Item appears in HA List (only Title)
- Bring maintains Details (no hash visible)

Other Case:
- Adding Item in HA
- Synced to bring -> In Bring Item has Hash.
- Changing Detail in Bring to... lets say: "200g"
- With next bring sync of HA Details will get overwritten and replaced with the hash.

So it appears that in the first Case: The ID is not necessary (or its weirdly working without the hash :D - I doubt that its using the "Details"-field as id (because I have 2 Items with "200g" in my list))
Maybe it helps?

**Comment by vlebourl** [comment]
> Hm... Cant we use the name+Detail from bring as id (should be unique in most cases) and bypass the uuid generation?

I'm actually working on that right now :)

> Also noticed this Case:
> 
> When adding an Item with details in Bring first
> Calling Bring-Sync Service in HA
> Item appears in HA List (only Title)
> Bring maintains Details (no hash visible)

that's because ATM the details in Bring are used as id in HA if it exists.

> Other Case:
> 
> Adding Item in HA
> Synced to bring -> In Bring Item has Hash.
> Changing Detail in Bring to... lets say: "200g"
> With next bring sync of HA Details will get overwritten and replaced with the hash

That's also expected in the current implementation as only the name is used to find an item in Bring's list, and the we push the id as detail.
.

**Comment by vlebourl** [comment]
I think the right way to do it would be to have a single dictionary that contains the HA hash as key and the item with all its components as value. And then, synchronizing would consist in updating the dict with elements from HA and Bring with a smart rule to identify corresponding items, and then updating both HA and Bring to match the dict.

**Comment by vlebourl** [comment]
I've added releases to make keeping track of installed versions easier. 
I've pushed a new beta with changes that should suit your needs. Let me know if it works!

0.2-beta4

**Comment by Camatobe** [comment]
Wow many thanks. I try to figure out how to go to the pre-release build in HACS. - I will test it in the evening.

**Comment by vlebourl** [comment]
open the reinstall menu and check the tick!
![image](https://user-images.githubusercontent.com/5701372/99149700-fb1b9600-268f-11eb-9333-54c7587e11a6.png)

**Comment by vlebourl** [comment]
you can add details from home asssitant with the brackets: in HA, `item [details]` will appear in Bring as 
```
item
details
```

**Comment by Camatobe** [comment]
Nice Work! This is working amazingly well!
Tested a wide range of cases. Adding via HA, Adding via Bring, syncing with multiple Items of the same title, but different detail. Works really really well. 

Some Warnings popped up. I don’t checked if they were there before.

I paste down for the sake of documenting:
`Logger: homeassistant.util.async_
Source: util/async_.py:105 
First occurred: 17:46:50 (32 occurrences) 
Last logged: 17:52:13

Detected I/O inside the event loop. This is causing stability issues. Please report issue to the custom component author for shopping_list doing I/O at custom_components/shopping_list/bring.py, line 166: cls._translations = requests.get(
Detected I/O inside the event loop. This is causing stability issues. Please report issue to the custom component author for shopping_list doing I/O at custom_components/shopping_list/bring.py, line 69: items = requests.get(
Detected I/O inside the event loop. This is causing stability issues. Please report issue to the custom component author for shopping_list doing I/O at custom_components/shopping_list/bring.py, line 104: requests.put(
Detected I/O inside the event loop. This is causing stability issues. Please report issue to the custom component author for shopping_list doing I/O at custom_components/shopping_list/bring.py, line 115: requests.put(
Detected I/O inside the event loop. This is causing stability issues. Please report issue to the custom component author for shopping_list doing I/O at custom_components/shopping_list/bring.py, line 93: requests.put( `

**Comment by vlebourl** [comment]
Yeah, those i/o warnings are expected. That's the next step, make it non blocking

**Comment by vlebourl** [comment]
I've published a new beta (0.3-beta2) that should fix those I/O warnings. Could you try it for me? cheers :)

**Comment by Camatobe** [comment]
Nicely done! Tested the .3 beta2 in various cases. (Adding Items back and forth, removing, adding back, syncing)
No log errors, no visual bugs! Very very nice!

I identified another "bug" or a possible feature request but will open another Issue for that :)

This Issue here can be left closed for sure.
